zkml/src/layers/concat_matmul.rs

```diff
@@ -10,78 +10,559 @@
+
+        Ok((
+            points[concat_dimension],
+            points[row_dimension],
+            points[col_dimension],
+        ))
+    }
+}
+
 use super::provable::LayerOut;
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ConcatMatMul {
-    permute: Option<Vec<usize>>,
+    permutations: MatrixPermutations,
     /// It tells what is the maximum bit size we ever expect the output of this layer to be.
     /// NOTE: This is a config item normally but we need this information during quantization.
     /// Best would be to rework quantization trait to include such config items.
     intermediate_bit_size: usize,
 }
 
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct ConcatMatMulCtx<E> {
+    pub(crate) node_id: NodeId,
+    pub(crate) matrix_poly_aux: VPAuxInfo<E>,
+    permutations: MatrixPermutations,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[serde(bound(serialize = "E: Serialize", deserialize = "E: DeserializeOwned"))]
+pub struct ConcatMatMulProof<E: ExtensionField> {
+    sumcheck_proof: IOPProof<E>,
+    /// The individual evaluations of the individual polynomial for the last random part of the
+    /// sumcheck. One for each polynomial involved in the "virtual poly".
+    /// Since we only support quadratic right now it's a flat list.
+    individual_claims: Vec<E>,
+}
+
+impl<E: ExtensionField> ConcatMatMulProof<E> {
+    /// Returns the individual claims f_1(r) f_2(r)  f_3(r) ... at the end of a sumcheck multiplied
+    /// together
+    pub fn individual_to_virtual_claim(&self) -> E {
+        self.individual_claims
+            .iter()
+            .fold(E::ONE, |acc, e| acc * *e)
+    }
+}
+
 const DEFAULT_INTERMEDIATE_BIT_SIZE: usize = 25;
 
 impl ConcatMatMul {
-    pub fn new() -> Self {
+    pub fn new(left: InputMatrixDimensions, right: InputMatrixDimensions) -> Self {
         Self {
-            permute: None,
+            permutations: MatrixPermutations {
+                left,
+                right,
+                permute: None,
+            },
             intermediate_bit_size: DEFAULT_INTERMEDIATE_BIT_SIZE,
         }
     }
-    pub fn new_with_permute(permutation: Vec<usize>) -> Self {
+    pub fn new_with_permute(
+        left: InputMatrixDimensions,
+        right: InputMatrixDimensions,
+        permutation: Permutation,
+    ) -> Self {
         Self {
-            permute: Some(permutation),
+            permutations: MatrixPermutations {
+                left,
+                right,
+                permute: Some(permutation),
+            },
             intermediate_bit_size: DEFAULT_INTERMEDIATE_BIT_SIZE,
         }
     }
-    pub fn with_max_shapes(self, max_shapes: Vec<Shape>) -> Self {
-        Self::ensure_shape_consistency(&max_shapes).unwrap();
-        let matrix_shape = max_shapes.into_iter().next().unwrap().slice(1..2);
-        let intermediate_bit_size = matrix_shape.matmul_output_bitsize();
+    pub fn with_max_shapes(
```

Comment:
Why is this function necessary and what does it do ? It's unclear why should we specify "maximum shapes", is it because of max sequence length and padding ?

Reply:
You actually introduced the method in a previous PR :P Btw, it is mostly employed to compute the `intermediate_bit_size` for quantization, so I renamed it and add a comment to explain its purpose in commit https://github.com/Lagrange-Labs/deep-prove/pull/183/commits/fef0d7bfcbc44242b1f0618a4243bfb583a5cbd8